rust/psibase/src/flags.rs

```diff
@@ -0,0 +1,143 @@
+use crate::check;
+
+pub trait HasFlags<T> {
+    fn set_flag(&mut self, index: u8, value: bool);
+    fn get_flag(&self, index: u8) -> bool;
+}
+
+/// # Flags
+///
+/// `Flags` represents a bitset structure with a boolean value for each bit.
+///
+/// Provides a convenient way to manage an integer that represents a set of boolean flags.
+///
+/// It can work with any unsigned integer type (`u8`, `u16`, `u32`, `u64`, etc.). For example, Flags<u8>
+/// can work with 8 flags, Flags<u16> can work with 16 flags, etc.
+///
+/// # Type Parameter
+///
+/// * `T` - The unsigned integer type to use for storing flags. Must implement
+///   the required bitwise operation traits.
+pub struct Flags<T> {
+    value: T,
+}
+
+impl<T> Flags<T>
+where
+    T: Copy
+        + std::ops::BitOr<Output = T>
+        + std::ops::BitAnd<Output = T>
+        + std::ops::Not<Output = T>
+        + std::ops::Shl<u8, Output = T>
+        + PartialEq
+        + From<u8>,
+{
+    pub fn new(value: T) -> Self {
+        Self { value }
+    }
+
+    fn validate_index(index: u8) {
+        check(index < std::mem::size_of::<T>() as u8 * 8, "invalid index");
+    }
+
+    pub fn set(&mut self, index: u8, value: bool) -> &mut Self {
+        Self::validate_index(index);
+        if value {
+            self.value = self.value | (T::from(1) << index);
+        } else {
+            self.value = self.value & !(T::from(1) << index);
+        }
+        self
+    }
+
+    pub fn get(&self, index: u8) -> bool {
+        Self::validate_index(index);
+        (self.value & (T::from(1) << index)) != T::from(0)
+    }
+
+    pub fn value(&self) -> T {
+        self.value
+    }
+}
+
+/// Macro to generate flag types with constants, get() method, and GraphQL schema
+///
+/// > **WARNING:** This macro does not check if the number of flags exceeds the storage
+/// >              type capacity.
+///
+/// ### Usage example
+/// ```rust,ignore
+/// define_flags!(StructName, u8, {
+///     first_flag,
+///     second_flag,
+///     third_flag,
+/// });
+/// ```
+///
+/// This generates:
+///
+/// - A struct with the given name and bool fields (snake_case names)
```

Comment:
We've been using camelCase for JSON. I believe this was originally because that's the graphQL convention. async_graphql handles the conversion automatically, but serde need `rename`.